processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingOptions.java

```diff
@@ -54,13 +58,171 @@ public MappingOptions(Map<String, List<Mapping>> mappings, IterableMapping itera
+                PropertyEntry property = first( first( mapping ).getTargetReference().getPropertyEntries() );
+                if ( !mappingsKeyedByProperty.containsKey( property ) ) {
+                    mappingsKeyedByProperty.put( property, new ArrayList<Mapping>() );
+                }
+                mappingsKeyedByProperty.get( property ).add( newMapping );
+            }
+        }
+
+        // now sort them into mapping options
+        Map<PropertyEntry, MappingOptions> result = new HashMap<PropertyEntry, MappingOptions>();
+        for (  Map.Entry<PropertyEntry, List<Mapping>> mappingKeyedByProperty : mappingsKeyedByProperty.entrySet() ) {
+            Map<String, List<Mapping>> newEntries = new HashMap<String, List<Mapping>>();
+            for ( Mapping newEntry : mappingKeyedByProperty.getValue() ) {
+                newEntries.put( newEntry.getTargetName(), Arrays.asList( newEntry ) );
+            }
+            result.put( mappingKeyedByProperty.getKey(), forMappingsOnly( newEntries ) );
+        }
+        return result;
+    }
+
+    /**
+     * Check there are nested target references for this mapping options.
+     *
+     * @return boolean, true if there are nested target references
+     */
+    public boolean hasNestedTargetReferences() {
+        boolean found = false;
+        Iterator<List<Mapping>> mapIterator = mappings.values().iterator();
+        while ( mapIterator.hasNext() && !found ) {
+            Iterator<Mapping> listIterator = mapIterator.next().iterator();
+            while ( listIterator.hasNext() && !found ) {
+                TargetReference targetReference = listIterator.next().getTargetReference();
+                found =  targetReference.isValid() && targetReference.getPropertyEntries().size() > 1;
+            }
+        }
+        return found;
+    }
+
+    /**
+     *
+     * @return all dependencies to other properties the contained mappings are dependent on
+     */
+    public List<String> collectNestedDependsOn() {
+
+        List<String> nestedDependsOn = new ArrayList<String>();
+        for ( List<Mapping> mappingList : mappings.values() ) {
+            for ( Mapping mapping : mappingList ) {
+                nestedDependsOn.addAll( mapping.getDependsOn() );
+            }
+        }
+        return nestedDependsOn;
+    }
+
+
+    /**
+     * Splits the MappingOptions into possibly more MappingOptions based on each source method parameter type.
+     *
+     * Note: this method is used for forging nested update methods. For that purpose, the same method with all
+     *       joined mappings should be generated. See also: NestedTargetPropertiesTest#shouldMapNestedComposedTarget
+     *
+     * @return the split mapping options.
+     *
+     */
+    public Map<Parameter, MappingOptions> splitOnSourceParameterType() {
+
+        Set<Parameter> parameters = new HashSet<Parameter>();
+        Map<Type, List<Mapping>> mappingsKeyedByParameterType = new HashMap<Type, List<Mapping>>();
```

Comment:
If we do this, then if 2 parameters have the same type like in the `ChartEntryToArtist` then we are going to return the same `MappingOptions` for `chart1` and `chart2`. Is this expected?